### ISSUE TITLE
[macOS] Canvas/Timeline shortcut issues

### ISSUE THREAD

**Issue by pgilfernandez** [friction2d/friction#324]
In some scenarios you want to duplicate keyframes and you end up duplicating layers/groups/paths as well.

For instance, if you select a square in the viewport and then you go to the keyframes timeline, select a keyframe, scrub the "time line" and duplicate (by menu or shortcut) the result is that you are duplicating both the path and creating a duplicate of the keyframe...

Following this example, if you select a square in the viewport (it gets selected in the timeline) but you first click in an empty space then you are kind of "unselecting all". Now you go back to the keyframes timeline, select a keyframe, scrub the "time line" and duplicate (by menu or shortcut) the result is now is correct, just the keyframe gets duplicated.

The proposed behavior here would be that if your mouse is in the timeline, if you run the "duplicate" command, if there is a keyframe selected the only thing that gets duplicated is the keyframe and not any other thing but keyframes. In case there is no keyframe selected then it could duplicate layers/groups/paths...

Could you reproduce this?

Thanks!

**Comment by pgilfernandez** [comment]
I got time to record a video:


https://github.com/user-attachments/assets/409df838-b2c6-4608-b951-832747b5db56

**Comment by KirbysDarkNebula** [comment]
I'm unable to reproduce this issue in 1.0.0-beta1, the only problem I found was that keyframe selection would not get removed after clicking on the viewport which leads to confusing behaviour 

Doing the same, getting no object + keyframe duplication, only keyframe duplication:
https://github.com/user-attachments/assets/fb5e5c23-14cb-4513-aa9d-300a702a8c4e

**Comment by pgilfernandez** [comment]
Then it only happens in macOS... @KirbysDarkNebula, what OS are you using?

**Comment by KirbysDarkNebula** [comment]
I'm using Linux with the Appimage release

**Comment by pgilfernandez** [comment]
> I'm using Linux with the Appimage release

Thanks!
OK, then it happens in macOS and I guess it will not happen on Windows neather...

**Comment by rodlie** [comment]
Seems like another awesome macOS issue :smile: 

> the only problem I found was that keyframe selection would not get removed after clicking on the viewport which leads to confusing behaviour

Keyframe selection has nothing to do with the viewport.

**Comment by pgilfernandez** [comment]
@rodlie could you point me out where to look at in order to try fixing this for macOS? I would like to try...

Thanks

**Comment by rodlie** [comment]
I have not looked at the issue yet, currently busy with work stuff.

I assume that the copy keyframe action does not reserve/block the key event and you end up with multiple actions on mac.

Will be more involved with the project this weekend (as I want to push for a last beta release next week).

**Comment by pgilfernandez** [comment]
Don't worry, I'll investigate. I wanted it just to get faster to the "target" and find a fix.

Nice to hear about a second beta. I hope you have time before to have a look at my PRs and I hope you introduce the experimental macOS build 🙏🏻😄 hehehe

**Comment by rodlie** [comment]
Been looking at this issue and it will probably not be a simple fix.

The custom class NoShortcutAction still register shortcuts on Mac, so that's why keys and objects are copied when doing cmd+c on the timeline (since cmd+c is set as shortcut on a NoShortcutAction :smile )

This is fixable, all NoShortcutAction's will need to be reviewed and verified that they works as they should on Mac, and what happens when we disable all shortcuts on them.

But, we have one more issue, duplicated copy.

When pressing cmd+c in the timeline you will notice that focus switches to the window, and that triggers a new copy action. For some reason focus change when pressing cmd+c on Mac, this is bad as Friction is focus based for shortcuts etc.

**Comment by pgilfernandez** [comment]
I understand, not easy fix.

I was trying to filter the clipboard so that just paste the type of element I wanted (keyframes for instance if the timeline was on focus) but if you say the focus changes when the key shortcut is triggered then it would never work...

**Comment by rodlie** [comment]
Fixed in https://github.com/friction2d/friction/commit/fdb445507f83e49cd8e040a62beb78f8175b7df2

Note that there might be more broken shortcuts on macOS.

**Comment by rodlie** [comment]
Possible fix: https://github.com/friction2d/friction/commit/0b8907b4bb38d3df9789b63fcfd7a5a0717bfe25

**Comment by pgilfernandez** [comment]
> Possible fix: https://github.com/friction2d/friction/commit/0b8907b4bb38d3df9789b63fcfd7a5a0717bfe25

I'll try it tomorrow 👏🏻

**Comment by rodlie** [comment]
I will merge https://github.com/friction2d/friction/commit/0b8907b4bb38d3df9789b63fcfd7a5a0717bfe25 today, ok?

**Comment by pgilfernandez** [comment]
> I will merge [0b8907b](https://github.com/friction2d/friction/commit/0b8907b4bb38d3df9789b63fcfd7a5a0717bfe25) today, ok?

I'm compiling it right now, give me 15 minutes 😉

**Comment by rodlie** [comment]
No rush :smile:

**Comment by pgilfernandez** [comment]
Tested!

If you click, that is, focus, on the canvas or extrictly the timeline area (green areas in the following screenshot) and you run Duplicate/Copy/Cut/Paste shortcut, it works as expected.

But if you try to focus on any object of the hierarchy and try to Duplicate/Copy/Cut/Paste it, it doesn't work... unless you click in the canvas and go back to the hierarchy, then any Duplicating/Copy/Cut/Paste works...

<img width="773" alt="areas" src="https://github.com/user-attachments/assets/1ab2996c-d0bf-49fc-90db-a380cff7e5d5">

Does it match with your tests?

**Comment by rodlie** [comment]
> But if you try to focus on any object of the hierarchy and try to Duplicate/Copy/Cut/Paste it, it doesn't work

Thanks for catching, I only tested canvas vs. keys on the timeline... meh, need to find a fix.

**Comment by rodlie** [comment]
yeah... this will require some work, as we can't use focus for shortcuts. 'keysview' (the timeline/graph) only support shortcuts for keys, not objects. Since we now block the canvas shortcuts (that supports the objects shortcuts) cmd+copy/paste/duplicate only works on keys. If we unblock the canvas we make duplicates etc.

Back to the drawing board 😄

**Comment by rodlie** [comment]
New fix: https://github.com/friction2d/friction/commit/646ea6b90d9726ff6d8ee6b4c4a0acfbafb68f68

**Comment by pgilfernandez** [comment]
It's much better now. It feels natural...

Anyway there is still something that doesn't work as expected, if you select an object and a keyframe:
- and "you are focused" in the canvas with the cursor in the canvas, zone 3 or 4, it only duplicates the object
(OK, it behaves as expected)
- and "you are focused" in the timeline with the cursor in the timeline, it only duplicates the keyframe
(OK, it behaves as expected)
- and "you are focused" in the timeline and move your cursor to zone 3, it only duplicates the keyframe.
The natural behavior would be to **duplicate just the object as you are in a hierarchy**...

<img width="773" alt="areas2" src="https://github.com/user-attachments/assets/95e6659c-8e0b-434c-94e4-5cfffe246dff">

**Comment by rodlie** [comment]
> and "you are focused" in the timeline and move your cursor to zone 3, it only duplicates the keyframe.
The natural behavior would be to duplicate just the object as you are in a hierarchy...

Yes, any key event in zone 3 goes to zone 2 (keysview) as far as I'm aware, so if any keys are selected in zone 2 they take the key event, else the event goes to zone 1 (and the select objects take the event).

This is probably the case for Windows and Linux also (on a Mac so can't test).

**Comment by pgilfernandez** [comment]
Yes, it makes sense...
Then, it would be necessary that zone 3 gets its own key event or that it could send them to zone 1 (not to zone 2 as it does now...)

What about trying this workaround, anytime you click in zone 3 (and not using shift modifier, that is, a normal click), it forces to unselect everything...

**Comment by rodlie** [comment]
> What about trying this workaround, anytime you click in zone 3 (and not using shift modifier, that is, a normal click), it forces to unselect everything...

That would be bad UX, I might want to adjust/browse something in zone 3 while having selected keys in zone 2.

Possible solution:

1. add a hover state in zone 3
2. set hover state on enter/leave events in zone 3
3. find a way to either capture key events (this might break something else) or share hover state between zone 3 and 2, then we just check zone 3 hover state in zone 2.

**Comment by pgilfernandez** [comment]
It sounds nice!

**Comment by rodlie** [comment]
![Screenshot 2024-12-11 at 17 54 46](https://github.com/user-attachments/assets/e1b822f6-55c4-4a31-b450-50e42472eec3)

Note that both "boxes" has the same behavior, so shortcuts on box2 will also end up where you have the focus.

So, we might need to forward all key events to the canvas from the boxes, else we might confuse users.

**Comment by rodlie** [comment]
Suggested UX fix: https://github.com/friction2d/friction/commit/4855bba981f93171c89fc96b8932be26e53bdc1a

**Comment by pgilfernandez** [comment]
> Suggested UX fix: [4855bba](https://github.com/friction2d/friction/commit/4855bba981f93171c89fc96b8932be26e53bdc1a)

Yes!
it works perfect for me, let me test deeper along this weekend but I think it is perfect now. What to you think?

**Comment by rodlie** [comment]
> What to you think?

Seems to work ok on my macbook, untested on Linux and Windows (will test later today).

Need to make sure we don't break anything (existing shortcuts etc).

**Comment by rodlie** [comment]
Yeah... This does not work on Linux.

On Linux/Windows all key events goes to processKeyEvent (in mainwindow) where the events goes to the active KFT target (canvas or timeline). On macOS we have done various workarounds that seems to work ok, but in reality we just bypass/ignore processKeyEvent it seems.

The custom key event handler from enve should be refactored, but it's not something I think is doable for v1.0.

**Comment by rodlie** [comment]
New branch for various macOS stuff that will be merged into main before rc1: https://github.com/friction2d/friction/tree/macos

I decided to keep the key event changes, but only for macOS.

This issue will now track the refactor of key events and focus for all platforms.
